nodejs/week3/src/backend/api/meals.js

```diff
@@ -0,0 +1,148 @@
+const express = require("express");
+const app = express();
+const router = express.Router();
+const knex = require("../database");
+
+//Get all meals
+router.get("/", async (req, res) => {
+  try {
+    const mealslist = await knex("meals").select("*");
+    res.send(mealslist);
+  } catch (error) {
+    throw error;
+  }
+});
+
+router.get("/titles", async (request, response) => {
+  try {
+    // knex syntax for selecting table records.
+    const titles = await knex("meals").select("title");
+    response.json(titles);
+  } catch (error) {
+    throw error;
+  }
+});
+
+//GET api/meals/ query parameters
+
+router.get("/", async function (req, res) {
+  const {
+    maxPrice,
+    title,
+    availableReservations,
+    createdAfter,
+    limit,
+  } = req.query;
+  const mealTitles = await knex.select("*").table("meals");
+
+  if (Object.keys(req.query).length === 0) {
+    res.send(mealTitles);
+  }
+
+  if (maxPrice) {
+    try {
+      const maxPriceMeals = await knex(meals)
+        .select("*")
+        .where("price", "<", maxPrice);
+      console.log(maxPriceMeals);
+      res.send(maxPriceMeals);
+    } catch (error) {
+      error;
+    }
+  } else if (title) {
+    try {
+      const titles = await knex(meals)
+        .select("*")
+        .where("title", "like", `%${title}%`);
+      console.log(titles);
+      res.send(titles);
+    } catch (error) {
+      error;
+    }
+  } else if (availableReservations === "true") {
+    try {
+      const availableReservationsMeal = await knex
+        .select("*")
+        .sum({
+          sumTotal: "reservations.number_of_guests",
+        })
+        .table("meals")
+        .join("reservations", {
+          "meals.id": "reservations.meal_id",
+        })
+        .groupBy("meals.title")
+        .having("sumTotal", ">", "Max_reservations");
+      console.log(availableReservationsMeal);
+      res.send(availableReservationsMeal);
+    } catch (error) {
+      error;
+    }
+  } else if (createdAfter) {
+    try {
+      const datecreated = await knex
+        .select("*")
+        .table("meals")
+        .where("created_date", "<", createdAfter);
+      console.log(datecreated);
+      res.send(datecreated);
+    } catch (error) {
+      error;
+    }
+  } else if (limit) {
+    try {
+      const limitMeals = await knex.select("*").table("meals").limit(limit);
+      console.log(limitMeals);
+      res.send(limitMeals);
+    } catch (error) {
+      error;
+    }
+  }
+});
```

Comment:
All these conditions should be applied to the same query, and not to 4 different queries. The all should be computed together and used to constrain a SELECT query to the database. E.g:

api/meals?maxPrice=90 => return all meals under 90 gold coins
api/meals?maxPrice=90&limit=3 => return at most 3 meals under 90 gold coins
api/meals?maxPrice=90&limit=3&availableReservations=true => return at most 3 meals under 90 gold coins with available reservations.

Try storing the query knex('meals') in a variable a successively apply where statements to it based on the query params, or sending more properties into the `where` method.